Improve floating point accuracy in percentile

wphicks opened the pull request:
Improve the floating point accuracy of the linear interpolation formula used in `percentile_weightnening` kernel of `cupy.percentile`
Resolve #4607

wphicks commented:
The fix here is relatively straightforward and exactly mirrors the fix used in numpy. The associated unit test is a little more annoying. In numpy, Hypothesis was used to create a more general test of this problem, but since it looks like we aren't using Hypothesis in cupy, I used Hypothesis to find a "magic value" that would produce the error and then just hardcoded that value into the test that I've submitted here. Does that seem reasonable or do folks have some other idea on how this might be tested?

leofang commented on cupy/_statistics/order.py:
```diff
@@ -245,10 +245,17 @@ def _quantile_unchecked(a, q, axis=None, out=None, interpolation='linear',
+
+            U diff = a[offset_top] - a[offset_bottom];
+
+            if (weight_above < 0.5) {
+                ret = a[offset_bottom] + diff * weight_above;
+            } else {
+                ret = a[offset_top] - diff * (1 - weight_above);
+            }
```
It looks like this branching would cause thread divergence. Could you benchmark against a few cases and report the timing before and after this patch? You can use `cupyx.time.repeat()` for a quick one-liner timing.

wphicks replied:
Good thing to check! Here are the results for master:
```
For array of length 5
f                   :    CPU:  280.502 us   +/- 3.425 (min:  274.539 / max:  320.130) us     GPU-0:  420.902 us   +/-18.950 (min:  410.592 / max:  524.288) us

For array of length 50
f                   :    CPU:  279.530 us   +/- 3.893 (min:  273.767 / max:  489.012) us     GPU-0:  415.905 us   +/- 2.777 (min:  358.080 / max:  539.712) us

For array of length 500
f                   :    CPU:  279.705 us   +/- 3.187 (min:  273.260 / max:  404.803) us     GPU-0:  415.872 us   +/- 2.557 (min:  411.648 / max:  536.448) us

For array of length 5000
f                   :    CPU:  279.608 us   +/- 3.561 (min:  274.327 / max:  459.135) us     GPU-0:  415.901 us   +/- 2.545 (min:  411.648 / max:  504.128) us

For array of length 50000
f                   :    CPU:  279.674 us   +/- 3.171 (min:  273.374 / max:  371.667) us     GPU-0:  415.424 us   +/- 2.578 (min:  410.208 / max:  528.384) us

For array of length 500000000
f                   :    CPU:  279.140 us   +/- 3.060 (min:  274.150 / max:  336.556) us     GPU-0:  415.554 us   +/- 2.390 (min:  411.648 / max:  468.448) us
```
and here are results for this PR:
```
For array of length 5
f                   :    CPU:  281.037 us   +/- 4.982 (min:  275.461 / max:  411.347) us     GPU-0:  418.918 us   +/-18.364 (min:  410.624 / max:  528.576) us

For array of length 50
f                   :    CPU:  278.693 us   +/- 4.214 (min:  273.581 / max:  479.576) us     GPU-0:  414.290 us   +/- 2.819 (min:  409.792 / max:  530.112) us

For array of length 500
f                   :    CPU:  280.481 us   +/-21.542 (min:  273.537 / max:  893.934) us     GPU-0:  415.439 us   +/-15.847 (min:  410.080 / max:  909.312) us

For array of length 5000
f                   :    CPU:  280.989 us   +/- 3.962 (min:  275.230 / max:  428.586) us     GPU-0:  416.624 us   +/- 2.904 (min:  357.152 / max:  553.696) us

For array of length 50000
f                   :    CPU:  281.072 us   +/- 3.541 (min:  275.484 / max:  429.908) us     GPU-0:  416.318 us   +/- 2.795 (min:  411.936 / max:  563.200) us

For array of length 500000000
f                   :    CPU:  281.569 us   +/- 3.258 (min:  275.810 / max:  379.555) us     GPU-0:  416.871 us   +/- 2.502 (min:  412.320 / max:  513.376) us
```
Performance seems pretty comparable, which is what I would expect given how lightweight the work in each branch is, but let me know if you want me to try any other benchmarks or more rigorous tests.

leofang replied:
Thanks! Sorry I dropped the ball. Is this done with a random array in [0, 1] with q=0.5?

wphicks replied:
Terribly sorry, when I went back to look at this, I realized there was a typo in my benchmarking script, so the size was not getting reset properly in between runs. The results are still quite comparable between master and the PR, though, even with the correction. I uploaded the script I used to [this gist](https://gist.github.com/wphicks/774701358b3aa4be33277aa4cc7cd7d2). It computes the 0, 50th, and 100th percentile for randomly generated arrays in [0, 1].

Here are the updated results:

master
```
For array of length 5
f                   :    CPU:  241.075 us   +/- 3.559 (min:  236.662 / max:  336.358) us     GPU-0:  246.124 us   +/- 3.660 (min:  240.576 / max:  342.400) us

For array of length 50
f                   :    CPU:  240.851 us   +/- 3.512 (min:  236.594 / max:  391.911) us     GPU-0:  245.895 us   +/- 3.601 (min:  241.600 / max:  397.312) us

For array of length 500
f                   :    CPU:  241.476 us   +/- 2.963 (min:  237.328 / max:  326.025) us     GPU-0:  246.539 us   +/- 3.059 (min:  241.888 / max:  332.320) us

For array of length 5000
f                   :    CPU:  279.828 us   +/- 3.995 (min:  274.140 / max:  402.477) us     GPU-0:  419.932 us   +/-18.860 (min:  356.352 / max:  518.144) us

For array of length 50000
f                   :    CPU:  279.626 us   +/- 4.327 (min:  273.721 / max:  405.583) us     GPU-0:  416.089 us   +/- 2.952 (min:  411.360 / max:  488.512) us

For array of length 5000000
f                   :    CPU:  284.788 us   +/-13.464 (min:  276.029 / max:  969.155) us     GPU-0: 2994.113 us   +/- 9.859 (min: 2926.592 / max: 3420.000) us
```

This PR
```
For array of length 5 
f                   :    CPU:  238.881 us   +/- 3.213 (min:  234.596 / max:  361.869) us     GPU-0:  243.750 us   +/- 3.312 (min:  238.624 / max:  368.832) us
                                                                                                       
For array of length 50                                                                                 
f                   :    CPU:  239.363 us   +/- 2.571 (min:  235.609 / max:  284.523) us     GPU-0:  244.241 us   +/- 2.673 (min:  239.648 / max:  290.816) us
                                                                                                       
For array of length 500                                                                                
f                   :    CPU:  240.826 us   +/- 3.010 (min:  236.789 / max:  316.428) us     GPU-0:  245.730 us   +/- 3.502 (min:  241.664 / max:  405.536) us
                                                                                                       
For array of length 5000                                                                               
f                   :    CPU:  276.527 us   +/- 3.105 (min:  271.654 / max:  317.652) us     GPU-0:  424.910 us   +/-15.261 (min:  417.792 / max:  517.696) us
                                                                                                       
For array of length 50000                                                                              
f                   :    CPU:  277.168 us   +/- 3.408 (min:  271.612 / max:  388.030) us     GPU-0:  422.038 us   +/- 2.751 (min:  417.824 / max:  530.432) us
                                                                                                       
For array of length 5000000                                                                            
f                   :    CPU:  280.512 us   +/- 3.707 (min:  273.256 / max:  356.554) us     GPU-0: 2993.909 us   +/- 5.918 (min: 2926.240 / max: 3227.808) us
```

leofang replied:
Nice, thanks!

leofang reviewed (commented):
> The fix here is relatively straightforward and exactly mirrors the fix used in numpy.

Any chance you could give a pointer to the Numpy counterpart for the record (either issue/PR or link to actual code is fine)?

> The associated unit test is a little more annoying. In numpy, Hypothesis was used to create a more general test of this problem, but since it looks like we aren't using Hypothesis in cupy, I used Hypothesis to find a "magic value" that would produce the error and then just hardcoded that value into the test that I've submitted here. Does that seem reasonable or do folks have some other idea on how this might be tested?

Well, even Hypothesis itself is fairly new to NumPy. It landed in NumPy only about a year ago (https://github.com/numpy/numpy/pull/15189) so don't feel annoyed that it's not adopted here 😄

Could you add a few more magic values here, and also test against NumPy's outputs (as done in all other tests throughout the codebase)? You could return the output in the test function, decorated by `@testing.numpy_cupy_allclose()`. This way we get a stronger confidence in the results.

wphicks commented:
> Any chance you could give a pointer to the Numpy counterpart for the record (either issue/PR or link to actual code is fine)?
Oh yes! Sorry about that; included it in the issue but forgot to repost it here. The numpy PR is here: https://github.com/numpy/numpy/pull/16273, and it addressed https://github.com/numpy/numpy/issues/14685.

> Well, even Hypothesis itself is fairly new to NumPy. It landed in NumPy only about a year ago ([numpy/numpy#15189](https://github.com/numpy/numpy/pull/15189)) so don't feel annoyed that it's not adopted here smile
Haha! No worries; I was more annoyed with myself that I hadn't found a more elegant test without Hypothesis...

> Could you add a few more magic values here, and also test against NumPy's outputs (as done in all other tests throughout the codebase)? You could return the output in the test function, decorated by `@testing.numpy_cupy_allclose()`. This way we get a stronger confidence in the results.
Absolutely! Will update momentarily.

wphicks commented:
All right, got the test updated and posted some benchmark data. Let me know if I can tweak anything further or if you'd like to see something more from that test.

asi1024 commented on tests/cupy_tests/statistics_tests/test_order.py:
```diff
@@ -148,6 +148,25 @@ def controlled_allocator(size):
+    # See gh-4607
+    # "Magic" values used in this test were empirically found to result in
+    # non-monotonicity for less accurate linear interpolation formulas
+    @testing.for_orders((-29, -53, -207, -16373, -99999,), name='magic_value')
```
`testing.for_orders` is a decorator to parameterize tests with order of ndarray. Could you use `testng.parameterize` instead?

wphicks replied:
Oh yes! Sorry about that. Just pushed a correction.

asi1024 commented:
Jenkins, test this please.

asi1024 commented:
@wphicks Could you check test failure?

wphicks commented:
Ah, interesting. It looks like numpy is violating the percentile monotonicity test in CI. I couldn't reproduce this with numpy==1.20.0, but earlier versions (which don't include https://github.com/numpy/numpy/pull/16273) unsurprisingly demonstrate this issue. I wasn't immediately able to determine what numpy version was being used in CI to confirm that this is definitely the problem, but it seems likely. Does someone with better familiarity of the CI system know of a good way to check that version?

In terms of a fix, we can either add a requirement to the `test` section of `setup.py` for `numpy>=1.20.0` or we can skip the test if we detect an old version of numpy. I'm going to push a commit that changes the test requirement, but if that seems like too heavy-handed of a change, we can go with the skip instead.

asi1024 commented:
I see! Could you use decorator `@testing.with_requires('numpy>=1.20')`?

wphicks commented:
> I see! Could you use decorator `@testing.with_requires('numpy>=1.20')`?

Done! Thank you. I reverted the change to setup.py and added the decorator.

wphicks commented:
Looks like Github actions are [having some issues right now](https://www.githubstatus.com/), which seems to be responsible for the latest failure.

jakirkham commented:
Probably from this ( https://www.githubstatus.com/incidents/5qdkkyg958vy ). Maybe try pushing an empty commit?

leofang commented:
Power-cycle (close and reopen) should restart it.

leofang commented:
Jenkins, test this please

jakirkham commented:
This shows up in the Windows CI log. Does cuTENSOR get installed on Windows typically? If so, is this just an issue of a download timing out or something?

```
FAILED tests/cupyx_tests/tools_tests/test_install_library.py::TestInstallLibrary::test_install_cutensor
```

https://ci.preferred.jp/cupy.experimental.win.cuda100/68442/

leofang commented:
There're still tons of errors in the Windows CI so hopefully we can address them all before v9 is out. For this specific case, it's tracked in #4601. The issue is cuTENSOR for Windows is packaged as `.exe` which cannot be extracted without installing 3rd party tools like `7zip`, so this part cannot be automated and requires users' action. On conda-forge we don't have this problem because `7zip` was made a dependency on Windows. 

@jakirkham Perhaps it's easier if you could request the cuTENSOR team to upload a different, machine-readable format? 🙂 At least cuDNN does not have this issue.

leofang commented:
(btw all tests in `test_order.py` passed on Windows.)

jakirkham commented:
Ok let's move the Windows discussion to that issue then 🙂

asi1024 commented:
The CI failure is unrelated to this PR. LGTM!
